// ///////////////////////////////////////////////////////////////////////////////////////
//  Copyright (c) 2018 Logan Barnes - All Rights Reserved
//  Geometry Visualisation Server
// ///////////////////////////////////////////////////////////////////////////////////////
syntax = "proto3";

package gvs.proto;

import "google/protobuf/wrappers.proto";

message Message {
    string identifier = 1;
    string contents = 2;
}

message UUID {
    bytes value = 1;
}

//message Vec3 {
//    float x = 1;
//    float y = 2;
//    float z = 3;
//}

//message Mat4 {
//    repeated float data = 1;
//}

enum GeometryFormat {
    POINTS = 0;
    LINES = 1;
    LINE_STRIP = 2;
    TRIANGLES = 3;
    TRIANGLE_STRIP = 4;
    TRIANGLE_FAN = 5;
}

message GeometryFormatValue {
    GeometryFormat value = 1;
}

//enum DisplayMode {
//    POSITION = 0;
//    NORMALS = 1;
//    TEX_COORDS = 2;
//    COLOR = 3;
//    TEXTURE = 4;
//    SIMPLE_SHADING = 5;
//    ADVANCED_SHADING = 6;
//    WHITE = 7;
//}

//message DisplayModeValue {
//    DisplayMode value = 1;
//}

message GeometryInfo3D {
    repeated float positions = 1;
    repeated float normals = 2;
    repeated float tex_coords = 3;
    repeated float colors = 4;
    repeated uint32 indices = 5;
}

//message DisplayInfo {
//    google.protobuf.StringValue readable_id = 1;
//    GeometryFormatValue geom_format = 2;
//    DisplayModeValue shader_display_mode = 3;
//    Mat4 transform = 4;
//    Vec3 color = 5;
//    google.protobuf.BoolValue visible = 6;
//    google.protobuf.FloatValue opacity = 7;
//}

message SceneItemInfo {
    UUID uuid = 1;
    GeometryInfo3D geometry_info = 2;
    // DisplayInfo display_info = 3;
    // UUID parent = 4;
    // repeated UUID children = 5;
}
